lib/SILOptimizer/Mandatory/AddressLowering.cpp

```diff
@@ -282,312 +546,656 @@ class OpaqueValueVisitor {
+          continue;
 
-      for (auto result : II.getResults())
         visitValue(result);
+      }
     }
   }
+  canonicalizeReturnValues();
 }
 
-/// Populate `indirectApplies` and insert this apply in `valueStorageMap` if
-/// the call's non-tuple result is returned indirectly.
-void OpaqueValueVisitor::visitApply(ApplySite applySite) {
+/// Populate `indirectApplies`.
+void OpaqueValueVisitor::checkForIndirectApply(FullApplySite applySite) {
   auto calleeConv = applySite.getSubstCalleeConv();
   unsigned calleeArgIdx = applySite.getCalleeArgIndexOfFirstAppliedArg();
   for (Operand &operand : applySite.getArgumentOperands()) {
     if (operand.get()->getType().isObject()) {
       auto argConv = calleeConv.getSILArgumentConvention(calleeArgIdx);
       if (argConv.isIndirectConvention()) {
         pass.indirectApplies.insert(applySite);
+        return;
       }
     }
     ++calleeArgIdx;
   }
-
-  if (applySite.getSubstCalleeType()->hasIndirectFormalResults()) {
+  if (applySite.getSubstCalleeType()->hasIndirectFormalResults())
     pass.indirectApplies.insert(applySite);
-    if (!applySite.getType().is<TupleType>())
-      pass.valueStorageMap.insertValue(cast<ApplyInst>(applySite));
-
-    return;
-  }
 }
 
-/// If `value` is address-only add it to the `valueStorageMap`.
+/// If `value` is address-only, add it to the `valueStorageMap`.
 void OpaqueValueVisitor::visitValue(SILValue value) {
-  if (value->getType().isObject()
-      && value->getType().isAddressOnly(*pass.F)) {
-    if (pass.valueStorageMap.contains(value)) {
-      assert(isa<SILFunctionArgument>(
-          pass.valueStorageMap.getStorage(value).storageAddress));
-      return;
-    }
-    pass.valueStorageMap.insertValue(value);
+  if (!value->getType().isObject()
+      || !value->getType().isAddressOnly(*pass.function)) {
+    return;
+  }
+  if (pass.valueStorageMap.contains(value)) {
+    // Function arguments are already mapped from loads.
+    assert(isa<SILFunctionArgument>(
+        pass.valueStorageMap.getStorage(value).storageAddress));
+    return;
   }
+  pass.valueStorageMap.insertValue(value);
 }
 
-//===----------------------------------------------------------------------===//
-// OpaqueStorageAllocation: Generate alloc_stack and address projections for all
-// abstract storage locations.
-//===----------------------------------------------------------------------===//
-
-namespace {
-/// Allocate storage on the stack for every opaque value defined in this
-/// function in RPO order. If the definition is an argument of this function,
-/// simply replace the function argument with an address representing the
-/// caller's storage.
-///
-/// TODO: shrink lifetimes by inserting alloc_stack at the dominance LCA and
-/// finding the lifetime boundary with a simple backward walk from uses.
-class OpaqueStorageAllocation {
-  AddressLoweringState &pass;
+// Canonicalize returned values.
```

Comment:
Can you add a sentence here what this does? E.g. "Make sure that the operand of the return instruction is a tuple which with no other uses than the return instruction"